### DISCUSSION TITLE
stateEstimate logging varies with a static CF

### DISCUSSION THREAD

**Discussion by a240160572** [USC-ACTLab/crazyswarm#681] [Q&A]
Hey, I am working on my cf with Vicon motion capture system. However, there is some problem with running even HelloWorld.

The cf is capable of realizing the test flight with flowdeck in [https://www.bitcraze.io/documentation/tutorials/getting-started-with-stem-drone-bundle/](url). Therefore, I would say that the cf and crazyradio are functioning. But together with crazyswarm and Vicon, the performance is wired. The cf drifts during take-off. Here is a logging from Vicon when I run HelloWorld:
<img src="https://user-images.githubusercontent.com/48674648/187526105-5f67ed58-be90-449f-a80a-5de79bbca869.png" width="300" height="150">

During my troubleshooting, I noticed that the stateEstimate log varies even when the cf is standing still on floor, as the plot below: 
<img src="https://user-images.githubusercontent.com/48674648/187523370-c426ae61-9cc1-44b8-8fe8-25c0b04b8511.png" width="300" height="150">
I do not know if the wired state estimation is related to my problem. But it is appreciated if someone can help with an explanation.

**Comment by jpreiss** [comment]
probably some communication issue, check the radio channels and addresses in your crazyflies.yaml.

**Reply by a240160572** [reply]
> probably some communication issue, check the radio channels and addresses in your crazyflies.yaml.

Here is my CF id: 'radio://0/80/2M/E7E7E7E703',
and my crazyflies.yaml setting:
`crazyflies:
- channel: 80
  id: 3
  initialPosition: [0, 0, 0.0]
  type: default`

I think these two are aligned, right?

**Comment by whoenig** [comment]
Perhaps you forgot to upgrade the firmware of your Crazyradio. The default firmware does not support broadcasts.

**Reply by a240160572** [reply]
> Perhaps you forgot to upgrade the firmware of your Crazyradio. The default firmware does not support broadcasts.

The Crazyradio is with the latest release 0.53.
Here is my result from `lsusb -d 1915:7777 -v | grep bcdDevice`:
`can't get device qualifier: Resource temporarily unavailable
can't get debug descriptor: Resource temporarily unavailable
  bcdDevice            0.53`.

The warning seems does not influence the communication given I am able to connect to the cf with cfclient or execute HelloWorld remotely.

**Comment by whoenig** [comment]
a correctly updated firmware shows version 0.99. Your version does not
support broadcasts.

**Reply by a240160572** [reply]
Thank you for the quick response!
You mean the `bcdDevice           99.55`, the master version of [https://github.com/bitcraze/crazyradio-firmware](url)
I can barely connect to my crazyflie with that version. The lastest release on that git is 0.53, I noticed.
The crazyflie is with firmware of `Build 0:8c606d1f8f89 (2022.05) CLEAN`

Update: The firmware of crazyradio has been updated to the crazyswarm prebuilt version. The stateEstimate logging is normal with only static cf, at least stays at zeros. But the HelloWorld is still not a success.

**Reply by whoenig** [reply]
Yes, the master version (which should be the same as the prebuilt version, when using the correct make flags). This one should work fine both with cfclient and crazyswarm (with the crazylie firmware you mentioned).

What's the problem with helloworld now?

**Reply by a240160572** [reply]
When I ran HelloWorld (or other file, like figure8) and logged `stateEstimate` and `ctrltarget`, the crazyflie took off and drift/crashed. 

The `stateEstimate` is always like: 
<img src="https://user-images.githubusercontent.com/48674648/187747046-10799d17-67f3-4fa2-aa88-bb7e50f5b04c.png" width="300" height="150">
while all `ctrltarget` stays at 0. 

I tested the logging with some manual movement and it did work
<img src="https://user-images.githubusercontent.com/48674648/187748772-364a2400-cb2d-4eed-a547-566384d22997.png" width="300" height="150">

I don't know if this warning matters:
`/home/lin/crazyswarm/ros_ws/src/crazyswarm/scripts/pycrazyswarm/crazyflie.py:590: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  cfg = yaml.load(crazyflies_yaml)`

**Reply by whoenig** [reply]
The warning doesn't matter.

Do you run the Crazyswarm in both experiments (flying vs. manual movements)? Also, did you remove the flow deck for your experiments with vicon? Which object tracking mode are you using?

**Reply by a240160572** [reply]
1. Logged both via ros, one with python file and one without.
2. Yes, the flow deck is removed.
3. I used MotionCapture and the crazyflie has its unique 4-marker arrangement. 
4. Some addition info in launchfile: `estimator: 2 # 1: complementary, 2: kalman` and `controller: 2 # 1: PID, 2: mellinger `

**Reply by whoenig** [reply]
Some people have more luck with the PID controller. Also, check if it helps to set the "https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/launch/hover_swarm.launch#L36" to True (which would indicate that there are axis flips in Vicon Tracker).

**Reply by a240160572** [reply]
Hey, I tried both suggestions and here is the result:
1. PID: 

**Log from crazyflie**, the same as before
<img src="https://user-images.githubusercontent.com/48674648/187920796-f1023e61-a2a2-4ba7-a580-6947058af579.png" width="500" height="250">

**Log from Vicon**, you can see the drifting on Y axis is insane
<img src="https://user-images.githubusercontent.com/48674648/187920800-e191afb7-3683-4586-8fcd-b98a87ec26e7.png" width="500" height="250">

2. `send_position_only: True`

**Log from crazyflie**, at least it is no longer static. But both plots I align the epoch to 1662035960 till 1662035961.5, the `stateEstimate` seems to have some lag. 
<img src="https://user-images.githubusercontent.com/48674648/187921488-03a1ccab-0787-4afa-9126-74bc3a13246d.png" width="500" height="250">
In addition, the log from crazyswarm is stopped early with error: `[ INFO] [1662037014.773707201]: [cf3] Takeoff
terminate called after throwing an instance of 'std::runtime_error'
  what():  Unknown rigid body!`

**Log from Vicon**, matches the real flight
<img src="https://user-images.githubusercontent.com/48674648/187924058-588ef12b-6275-417d-a031-0e337f7170ba.png" width="500" height="250">

**Full track of the flight from Vicon**
<img src="https://user-images.githubusercontent.com/48674648/187925206-ec851e3b-b549-4b3b-ace1-b73de413e0ec.png" width="500" height="250">

In addition, even though the cf is drifting, `ctrltarget` does not respond to that and correct the motion.

**Comment by a240160572** [comment]
Dear @jpreiss and @whoenig, sorry for disturbing you. 

Is there a way I can have access to the data which is received by cf from vicon?
I was reading the vicon data via viconbridge package, and I would like to also inspect it from the cf side. Thank you.

**Reply by whoenig** [reply]
Yes, you can use the usd card. Whenever VICON information is received, an event is triggered (see https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/estimator.c#L190-L198 for the names).

**Reply by a240160572** [reply]
Thank you for the reply!
I tested with a manual 3-axis movement. 
The dot-line is the log from `viconbridge` package, while the star-line is the log from `locSrv.x/y/z`.
<img src="https://user-images.githubusercontent.com/48674648/191002696-807aff15-fba9-446d-ace9-a75467b2c696.png" width="500" height="250">


It is evident that there is a delay (about 1.5-2 seconds) with the data set. Any idea what the cause would be?

The delay remains with or without set `send_position_only` to `True`. When `send_position_only: False`, the log term `ext_pos.X/Y/Z` are always 0.
The delay also remains with a new CF and a new Crazyradio.

**Reply by whoenig** [reply]
I assume in this case you are streaming the logging data over the radio, rather than using a usd-card?

**Reply by a240160572** [reply]
Yes. Unfortunately, we do not have the SD card deck yet.

**Reply by whoenig** [reply]
That's a neat test. I am not sure what delay we should expect here. The logging messages are buffered on-board, so if the buffer is always full, it will take a while before you get the latest information. You could try reducing the frequency (to make the buffer less full). I also assume you only connect to a single CF in this case? Otherwise the bandwidth limitations create latency also.

Note that logging messages also contain an on-board timestamp (in milliseconds). So one could try to find the offset between the PC and on-board clock and then compare the values without delay in postprocessing.

**Reply by a240160572** [reply]
Yes, only one CF.
I tested with only one logging parameter `locSrv.x` with 10Hz. The same latency is observed as logging 12 parameters with 10Hz.
If we assume the misalignment is from the onboard clock/transmission latency and all data should be synchronized onboard, then what could be the reason for the failed HelloWorld test?

Here is the flight of my HelloWorld test.
https://drive.google.com/file/d/13Cbl62bNR2gjQ7h8gtKHeX7nxPBWX0jt/view?usp=sharing
You can notice the serious drift and at the end, it left the detection area of Vicon (marked in blue on the floor) and landed on the safety net. 

Thank you so much for all your help!

**Reply by whoenig** [reply]
In the video it really looks like the state estimator is not working properly - this sort of flight would also happen if you fly without any motion capture, just relying on on-board data. Could you look at the logging variables "stateEstimate.{x,y,z}", rather than "locSrv.{x,y,z}"?
Also, I assume you build the Crazyswarm with the Release flag? (https://github.com/USC-ACTLab/crazyswarm/blob/master/build.sh#L27).

**Reply by a240160572** [reply]
> Thank you for the reply! I tested with a manual 3-axis movement. The dot-line is the log from `viconbridge` package, while the star-line is the log from `locSrv.x/y/z`. <img alt="" width="500" height="250" src="https://user-images.githubusercontent.com/48674648/191002696-807aff15-fba9-446d-ace9-a75467b2c696.png">
> 
> It is evident that there is a delay (about 1.5-2 seconds) with the data set. Any idea what the cause would be?
> 
> The delay remains with or without set `send_position_only` to `True`. When `send_position_only: False`, the log term `ext_pos.X/Y/Z` are always 0. The delay also remains with a new CF and a new Crazyradio.

The `stateEstimate.{x,y,z}` is the same as `locSrv.{x,y,z}` in this plot.
The build.sh is also as default, `catkin_make -DCMAKE_BUILD_TYPE=RelWithDebInfo`.
